compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/ImmutableTypeCloner.java

```diff
@@ -702,10 +667,100 @@ private static BIntersectionType defineImmutableUnionType(Location pos, Types ty
+                                                                     PackageID pkgId, BSymbol owner,
+                                                                     SymbolTable symTable,
+                                                                     BLangAnonymousModelHelper anonymousModelHelper,
+                                                                     Names names, Set<BType> unresolvedTypes,
+                                                                     BUnionType type, BType originalType) {
+        BTypeSymbol origBuiltInUnionTypeSymbol = type.tsymbol;
+
+        LinkedHashSet<BType> originalMemberList = type.getMemberTypes();
+        Optional<BIntersectionType> immutableTypeOptional = Types.getImmutableType(symTable, pkgId, type);
+        if (immutableTypeOptional.isPresent()) {
+            return immutableTypeOptional.get();
+        }
+
+        if (type.tag == TypeTags.JSON) {
+            defineImmutableJsonType(env, pkgId, owner, symTable, names, (BJSONType) type, originalType);
+        } else {
+            defineImmutableAnydataType(env, pkgId, owner, symTable, names, (BAnydataType) type, originalType);
+        }
+
+        return handleImmutableUnionType(pos, types, env, pkgId, owner, symTable, anonymousModelHelper, names,
+                                        unresolvedTypes, type, origBuiltInUnionTypeSymbol, originalMemberList);
+    }
+
+    private static void defineImmutableAnydataType(SymbolEnv env, PackageID pkgId, BSymbol owner, SymbolTable symTable,
+                                                   Names names, BAnydataType type, BType originalType) {
+        BTypeSymbol immutableAnydataTSymbol = getReadonlyTSymbol(names, type.tsymbol, env, pkgId, owner);
+
+        BAnydataType immutableAnydataType;
+        if (immutableAnydataTSymbol != null) {
+            immutableAnydataType =
+                    new BAnydataType(immutableAnydataTSymbol,
+                                     immutableAnydataTSymbol.name, type.flags | Flags.READONLY,
+                                     type.isNullable());
+            immutableAnydataTSymbol.type = immutableAnydataType;
+        } else {
+            immutableAnydataType =
+                    new BAnydataType(immutableAnydataTSymbol,
+                                     getImmutableTypeName(names, TypeKind.ANYDATA.typeName()),
+                                     type.flags | Flags.READONLY, type.isNullable());
+        }
+
+        BIntersectionType immutableAnydataIntersectionType =
+                createImmutableIntersectionType(pkgId, owner, originalType, immutableAnydataType, symTable);
+        Types.addImmutableType(symTable, pkgId, type, immutableAnydataIntersectionType);
+    }
+
+    private static void defineImmutableJsonType(SymbolEnv env, PackageID pkgId, BSymbol owner, SymbolTable symTable,
+                                                Names names, BJSONType type, BType originalType) {
```

Comment:
Can't we refactor the functions 'defineImmutableAnydataType' and 'defineImmutableJsonType' since they seem to have the same logic?

Reply:
`defineImmutableBuiltInUnionType` has the common logic. Will move the `Types.addImmutableType` bit also.